presto-password-authenticators/src/main/java/com/facebook/presto/password/file/PasswordStore.java

```diff
@@ -0,0 +1,129 @@
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.password.file.EncryptionUtil.doesBCryptPasswordMatch;
+import static com.facebook.presto.password.file.EncryptionUtil.doesPBKDF2PasswordMatch;
+import static com.facebook.presto.password.file.EncryptionUtil.getHashingAlgorithm;
+import static com.facebook.presto.spi.StandardErrorCode.CONFIGURATION_INVALID;
+import static com.facebook.presto.spi.StandardErrorCode.CONFIGURATION_UNAVAILABLE;
+import static java.lang.String.format;
+
+public class PasswordStore
+{
+    private static final Splitter LINE_SPLITTER = Splitter.on(":").limit(2);
+
+    private final Map<String, HashedPassword> credentials;
+    private final LoadingCache<Credential, Boolean> cache;
+
+    public PasswordStore(File file, int cacheMaxSize)
+    {
+        this(readPasswordFile(file), cacheMaxSize);
+    }
+
+    @VisibleForTesting
+    public PasswordStore(List<String> lines, int cacheMaxSize)
+    {
+        credentials = loadPasswordFile(lines);
+        cache = CacheBuilder.newBuilder()
+                .maximumSize(cacheMaxSize)
+                .build(CacheLoader.from(this::matches));
+    }
+
+    public boolean authenticate(String user, String password)
+    {
+        return cache.getUnchecked(new Credential(user, password));
+    }
+
+    private boolean matches(Credential credential)
+    {
+        HashedPassword hashed = credentials.get(credential.getUser());
+        return (hashed != null) && hashed.matches(credential.getPassword());
+    }
+
+    private static Map<String, HashedPassword> loadPasswordFile(List<String> lines)
+    {
+        Map<String, HashedPassword> users = new HashMap<>();
+        for (int lineNumber = 1; lineNumber <= lines.size(); lineNumber++) {
+            String line = lines.get(lineNumber - 1).trim();
+            if (line.isEmpty()) {
+                continue;
+            }
+
+            List<String> parts = LINE_SPLITTER.splitToList(line);
+            if (parts.size() != 2) {
+                throw invalidFile(lineNumber, "Expected two parts for user and password", null);
+            }
+            String user = parts.get(0);
+            String password = parts.get(1);
+
+            try {
+                if (users.put(user, getHashedPassword(password)) != null) {
+                    throw invalidFile(lineNumber, "Duplicate user: " + user, null);
+                }
+            }
+            catch (HashedPasswordException e) {
+                throw invalidFile(lineNumber, e.getMessage(), e);
+            }
+        }
+        return ImmutableMap.copyOf(users);
+    }
+
+    private static RuntimeException invalidFile(int lineNumber, String message, Throwable cause)
+    {
+        return new PrestoException(CONFIGURATION_INVALID, format("Error in password file line %s: %s", lineNumber, message), cause);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: lineNumber (declared at line 99). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 101. Consider formatting this data with &#39;%d&#39; instead.